File: src/pdftoolkit/ranges.py
"""Parsing of human-friendly, 1-based page selectors into 0-based indices.

Supported selector syntax (all 1-based, inclusive):
    "3"          -> a single page
    "2-4"        -> a contiguous, inclusive range
    "1,3,5"      -> a comma-separated list
    "1-3,5,7-8"  -> any combination of the above

Whitespace around tokens and commas is ignored. Reversed ranges such as
"4-2" are accepted and expand descending (4, 3, 2), which is what a user who
types them almost always means (e.g. for `reorder`).
"""

from __future__ import annotations

from typing import List


class RangeError(ValueError):
    """Raised when a page selector string cannot be parsed."""


def parse_pages(selector: str, page_count: int) -> List[int]:
    """Parse a 1-based page selector into a list of 0-based page indices.

    Args:
        selector: e.g. "1-3,5". Must be non-empty.
        page_count: total number of pages in the document, used for bounds
            checking and to honour the convention that omitting nothing.

    Returns:
        A list of 0-based indices in the order the user specified them.
        Duplicates are preserved (so "1,1" yields [0, 0]); callers that need
        uniqueness should de-duplicate themselves.

    Raises:
        RangeError: on empty input, malformed tokens, non-integer values,
            zero/negative page numbers, or numbers exceeding ``page_count``.
    """
    if selector is None:
        raise RangeError("no pages specified")
    text = selector.strip()
    if not text:
        raise RangeError("no pages specified")
    if page_count < 1:
        raise RangeError("document has no pages")

    indices: List[int] = []
    for raw_token in text.split(","):
        token = raw_token.strip()
        if not token:
            raise RangeError(f"empty page token in {selector!r}")
        if "-" in token:
            indices.extend(_parse_span(token, page_count, selector))
        else:
            indices.append(_parse_single(token, page_count, selector) - 1)
    return indices


def _parse_single(token: str, page_count: int, selector: str) -> int:
    """Parse a bare integer token into a validated 1-based page number."""
    try:
        value = int(token)
    except ValueError as exc:
        raise RangeError(f"invalid page number {token!r} in {selector!r}") from exc
    if value < 1:
        raise RangeError(f"page numbers start at 1, got {value} in {selector!r}")
    if value > page_count:
        raise RangeError(
            f"page {value} out of range (document has {page_count} page"
            f"{'s' if page_count != 1 else ''})"
        )
    return value


def _parse_span(token: str, page_count: int, selector: str) -> List[int]:
    """Parse an "a-b" span into a list of 0-based indices (inclusive)."""
    parts = token.split("-")
    if len(parts) != 2:
        raise RangeError(f"invalid range {token!r} in {selector!r}")
    start_s, end_s = parts[0].strip(), parts[1].strip()
    if not start_s or not end_s:
        raise RangeError(
            f"open-ended range {token!r} in {selector!r}; "
            f"both ends are required (e.g. '2-4')"
        )
    start = _parse_single(start_s, page_count, selector)
    end = _parse_single(end_s, page_count, selector)
    if start <= end:
        return [i - 1 for i in range(start, end + 1)]
    # Descending range, e.g. "4-2" -> [3, 2, 1].
    return [i - 1 for i in range(start, end - 1, -1)]

File: docs/js/ranges.js
/* Parsing of human-friendly, 1-based page selectors into 0-based indices.
 *
 * This is a faithful port of the CLI's `pdftoolkit.ranges` module so the web
 * app and the command line accept exactly the same syntax:
 *
 *   "3"          a single page
 *   "2-4"        a contiguous, inclusive range
 *   "1,3,5"      a comma-separated list
 *   "1-3,5,7-8"  any combination of the above
 *
 * Whitespace around tokens and commas is ignored. Reversed ranges such as
 * "4-2" are accepted and expand descending (4, 3, 2), which is what a user
 * who types them almost always means (e.g. for reorder).
 */

export class RangeError_ extends Error {
  constructor(message) {
    super(message);
    this.name = "RangeError";
  }
}

function parseSingle(token, pageCount, selector) {
  if (!/^\d+$/.test(token)) {
    throw new RangeError_(`invalid page number "${token}" in "${selector}"`);
  }
  const value = parseInt(token, 10);
  if (value < 1) {
    throw new RangeError_(`page numbers start at 1, got ${value} in "${selector}"`);
  }
  if (value > pageCount) {
    throw new RangeError_(
      `page ${value} out of range (document has ${pageCount} page` +
        `${pageCount !== 1 ? "s" : ""})`
    );
  }
  return value;
}

function parseSpan(token, pageCount, selector) {
  const parts = token.split("-");
  if (parts.length !== 2) {
    throw new RangeError_(`invalid range "${token}" in "${selector}"`);
  }
  const startS = parts[0].trim();
  const endS = parts[1].trim();
  if (!startS || !endS) {
    throw new RangeError_(
      `open-ended range "${token}" in "${selector}"; both ends are required (e.g. "2-4")`
    );
  }
  const start = parseSingle(startS, pageCount, selector);
  const end = parseSingle(endS, pageCount, selector);
  const indices = [];
  if (start <= end) {
    for (let i = start; i <= end; i++) indices.push(i - 1);
  } else {
    for (let i = start; i >= end; i--) indices.push(i - 1);
  }
  return indices;
}

/**
 * Parse a 1-based page selector into an array of 0-based page indices.
 *
 * Duplicates are preserved (so "1,1" yields [0, 0]); callers that need
 * uniqueness should de-duplicate themselves. Throws RangeError_ on empty
 * input, malformed tokens, zero/negative page numbers, or numbers exceeding
 * pageCount.
 */
export function parsePages(selector, pageCount) {
  if (selector === null || selector === undefined) {
    throw new RangeError_("no pages specified");
  }
  const text = String(selector).trim();
  if (!text) {
    throw new RangeError_("no pages specified");
  }
  if (pageCount < 1) {
    throw new RangeError_("document has no pages");
  }

  const indices = [];
  for (const rawToken of text.split(",")) {
    const token = rawToken.trim();
    if (!token) {
      throw new RangeError_(`empty page token in "${selector}"`);
    }
    if (token.includes("-")) {
      indices.push(...parseSpan(token, pageCount, selector));
    } else {
      indices.push(parseSingle(token, pageCount, selector) - 1);
    }
  }
  return indices;
}
